src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+    var styleUrl = self.styleUrl = convertStyleUrl(opts.style);
+
+    var map = self.map = new mapboxgl.Map({
+        container: self.div,
+
+        style: styleUrl,
+        center: convertCenter(opts.center),
+        zoom: opts.zoom,
+        bearing: opts.bearing,
+        pitch: opts.pitch,
+
+        interactive: !self.isStatic,
+        preserveDrawingBuffer: self.isStatic
+    });
+
+    self.rejectOnError(reject);
+
+    map.once('load', function() {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    });
+
+    // keep track of pan / zoom in user layout
+    map.on('move', function() {
+        var center = map.getCenter();
+        opts._input.center = opts.center = { lon: center.lng, lat: center.lat };
+        opts._input.zoom = opts.zoom = map.getZoom();
+    });
+
+    map.on('mousemove', function(evt) {
+        var bb = self.div.getBoundingClientRect();
+
+        // some hackery to get Fx.hover to work
+
+        evt.clientX = evt.point.x + bb.left;
+        evt.clientY = evt.point.y + bb.top;
+
+        evt.target.getBoundingClientRect = function() { return bb; };
+
+        self.xaxis.p2c = function() { return evt.lngLat.lng; };
+        self.yaxis.p2c = function() { return evt.lngLat.lat; };
+
+        Fx.hover(gd, evt, self.id);
+    });
+
+    map.on('click', function() {
+        Fx.click(gd, { target: true });
+    });
+
+    function unhover() {
+        Fx.loneUnhover(fullLayout._toppaper);
+    }
+
+    map.on('dragstart', unhover);
+    map.on('zoomstart', unhover);
+
+};
+
+proto.updateMap = function(calcData, fullLayout, resolve, reject) {
+    var self = this,
+        map = self.map;
+
+    self.rejectOnError(reject);
+
+    var styleUrl = convertStyleUrl(self.opts.style);
+
+    if(self.styleUrl !== styleUrl) {
+        self.styleUrl = styleUrl;
+        map.setStyle(styleUrl);
+
+        map.style.once('load', function() {
+
+            // need to rebuild trace layers on reload
+            // to avoid 'lost event' errors
+            self.traceHash = {};
+
+            self.updateData(calcData);
+            self.updateLayout(fullLayout);
+
+            self.resolveOnRender(resolve);
+        });
+    }
+    else {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    }
+};
+
+proto.updateData = function(calcData) {
+    var traceHash = this.traceHash;
+
+    var traceObj, trace, i, j;
+
+    // update or create trace objects
+    for(i = 0; i < calcData.length; i++) {
+        var calcTrace = calcData[i];
+
+        trace = calcTrace[0].trace;
+        traceObj = traceHash[trace.uid];
+
+        if(traceObj) traceObj.update(calcTrace);
+        else {
+            traceHash[trace.uid] = trace._module.plot(this, calcTrace);
+        }
+    }
+
+    // remove empty trace objects
+    var ids = Object.keys(traceHash);
+    id_loop:
```

Comment:
:cow2: I've never been a big fan of this - doesn't seem to be necessary in this instance either.